### PR TITLE
Wasm wasi

### PR DESCRIPTION
**Pull request by igoriakovlev**
This is `Wasm-WASI` target support for coroutines.

This is based on self-written `EventLoop` which has a simple implementation for queued line of callbacks with timeouts.

This allows to schedule tasks in scheduler. The `EventLoop` starts on the program entry-point exit (that supported by the Kotlin from 1.9.23).

The coroutine tests runner implemented with a direct `EventLoop` call so that _looks like_ the `RunBlocking` in `Native` or `JVM` implementation. We do not `RunBlocking` itself because there is no design for it in single-threaded environments.

The `EventLoop` throws an exception `EventLoopException` which is discussable because it is not supposed to be catched by user in normal code (but still possible for the `runTest` calls).

### CODE REVIEW

File: buildSrc/src/main/kotlin/kotlin-multiplatform-conventions.gradle.kts [hunk 1]
```diff
@@ -63,6 +63,18 @@ kotlin {
             api("org.jetbrains.kotlinx:atomicfu-wasm-js:${version("atomicfu")}")
         }
     }
+    @OptIn(org.jetbrains.kotlin.gradle.targets.js.dsl.ExperimentalWasmDsl::class)
+    wasmWasi {
+        nodejs()
+        compilations["main"]?.dependencies {
+            api("org.jetbrains.kotlinx:atomicfu-wasm-wasi:${version("atomicfu")}")
+        }
+        compilations.configureEach {
```
Comment (igoriakovlev) [inline buildSrc/src/main/kotlin/kotlin-multiplatform-conventions.gradle.kts:72]: Unfortunately we can't do this in `configure-compilation-conventions.gradle.kts` because there is no way to distingish `WasmWasi` target from `WasmJs` or `JS` targets there.

**Review by dkhalanskyjb — CHANGES_REQUESTED** [review]

**Comment by dkhalanskyjb** [comment]
Please change the base of the PR to `develop`.

File: buildSrc/src/main/kotlin/kotlin-multiplatform-conventions.gradle.kts [hunk 2]
```diff
@@ -101,15 +113,27 @@ kotlin {
...
+        val wasmWasiMain by getting {
+        }
+        val wasmWasiTest by getting {
+            dependencies {
+                api("org.jetbrains.kotlin:kotlin-test-wasm-wasi:${version("kotlin")}")
+            }
+        }
+        groupSourceSets("jsAndWasmShared", listOf("wasmWasi"), listOf("common"))
+        groupSourceSets("jsAndWasmJsShared", listOf("js", "wasmJs"), listOf("jsAndWasmShared"))
+        groupSourceSets("jsAndWasmShared", listOf("jsAndWasmJsShared"), emptyList(), alreadyExist = true)
```
[the remark this answers is not in the document]
Comment (igoriakovlev) [reply buildSrc/src/main/kotlin/kotlin-multiplatform-conventions.gradle.kts:125]: You cannot do the first line because `jsAndWasmShared` created in a second and vice-versa.

File: kotlinx-coroutines-core/wasmWasi/src/WasiDispatcher.kt [hunk 3]
```diff
@@ -0,0 +1,31 @@
+/*
+ * Copyright 2016-2021 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.coroutines
+
+import kotlinx.coroutines.internal.*
+import kotlin.coroutines.*
+
+internal object WasiDispatcher: CoroutineDispatcher(), Delay {
```
[the remark this answers is not in the document]
Comment (igoriakovlev) [reply kotlinx-coroutines-core/wasmWasi/src/WasiDispatcher.kt:10]: Well, because `WasiDispatcher` is not a `CoroutineContext`. But if you would like I will move it.

File: kotlinx-coroutines-core/wasmWasi/src/internal/CoroutineExceptionHandlerImpl.kt [hunk 4]
```diff
@@ -0,0 +1,9 @@
+/*
+ * Copyright 2016-2024 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.coroutines.internal
+
+internal actual fun propagateExceptionFinalResort(exception: Throwable) {
+    println(exception.toString())
```
[the remark this answers is not in the document]
Comment (igoriakovlev) [reply kotlinx-coroutines-core/wasmWasi/src/internal/CoroutineExceptionHandlerImpl.kt:8]: Well, I put the code which prints an error into error stream and then exit with not 0 error code. I could even send a signal if you would like to have most scariest termination. Please notice - this way all finalisers will not work.

File: kotlinx-coroutines-core/wasmWasi/src/internal/CoroutineRunner.kt [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+ * Copyright 2016-2024 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.coroutines.internal
+
+import kotlinx.coroutines.*
+import kotlin.coroutines.*
+
+@InternalCoroutinesApi
+public fun runTestCoroutine(context: CoroutineContext, block: suspend CoroutineScope.() -> Unit) {
```
[the remark this answers is not in the document]
Comment (igoriakovlev) [reply kotlinx-coroutines-core/wasmWasi/src/internal/CoroutineRunner.kt:11]: I prefer not to suppress anything. But if all reviewers ask for this then will do.

File: kotlinx-coroutines-core/wasmWasi/src/internal/EventLoop.kt [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+
+@WasmImport("wasi_snapshot_preview1", "poll_oneoff")
+private external fun wasiPollOneOff(ptrToSubscription: Int, eventPtr: Int, nsubscriptions: Int, resultPtr: Int): Int
+
+@WasmImport("wasi_snapshot_preview1", "clock_time_get")
+private external fun wasiRawClockTimeGet(clockId: Int, precision: Long, resultPtr: Int): Int
+
+private const val CLOCKID_MONOTONIC = 1
+
+internal class Event internal constructor(internal var callback: (() -> Unit)?, internal val absoluteTimeout: Long) {
```
[the remark this answers is not in the document]
Comment (igoriakovlev) [reply kotlinx-coroutines-core/wasmWasi/src/internal/EventLoop.kt:23]: It used for cancelling in the `WasiDispatcher`

File: kotlinx-coroutines-core/wasmWasi/src/internal/EventLoop.kt [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+@OptIn(UnsafeWasmMemoryApi::class)
+internal fun runEventLoop() {
+    if (nextCycleEvents.isEmpty()) return
+
+    withScopedMemoryAllocator { allocator ->
+        val ptrTo32Bytes = allocator.allocate(32)
+        val ptrTo8Bytes = allocator.allocate(8)
+        val ptrToSubscription = initializeSubscriptionPtr(allocator)
+
+        while (nextCycleEvents.isNotEmpty()) {
```
[the remark this answers is not in the document]
Comment (igoriakovlev) [reply kotlinx-coroutines-core/wasmWasi/src/internal/EventLoop.kt:147]: > If I understand correctly, here's how the event loop implementation works:
Is this a requirement for the implementation? Because there is no tests checking this. And I am not sure that the other implementations, like in `JS`, are following this.

File: buildSrc/src/main/kotlin/kotlin-multiplatform-conventions.gradle.kts (hunk 2 above)
[the remark this answers is not in the document]
Comment (dkhalanskyjb) [reply buildSrc/src/main/kotlin/kotlin-multiplatform-conventions.gradle.kts:125]: Yeah, sorry, I meant
```
        groupSourceSets("jsAndWasmJsShared", listOf("js", "wasmJs"), emptyList())
        groupSourceSets("jsAndWasmShared", listOf("jsAndWasmJsShared", "wasmWasi"), listOf("common"))
```

File: kotlinx-coroutines-core/wasmWasi/src/internal/EventLoop.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (dkhalanskyjb) [reply kotlinx-coroutines-core/wasmWasi/src/internal/EventLoop.kt:20]: `Event` can implement the `DisposableHandle` interface for this purpose.

File: kotlinx-coroutines-core/wasmWasi/src/internal/EventLoop.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (dkhalanskyjb) [reply kotlinx-coroutines-core/wasmWasi/src/internal/EventLoop.kt:148]: Yes, this is a requirement. Tests don't check this because, in multithreaded environments, we actually violate this intentionally sometimes. On JS, we don't use our own event loop, we rely on the one provided by the environment.

Now that I look at the implementation more closely, I don't understand the reason to implement new event loop logic at all. We can probably inherit from `EventLoopImplBase`. For that, we need to implement `EventLoopImplPlatform` for Wasm/WASI properly. If we go down that path, then, to implement `runEventLoop`, we can duplicate the logic from https://github.com/Kotlin/kotlinx.coroutines/blob/f9a4545e5d759595153dd089013cad141c96555b/kotlinx-coroutines-core/native/src/Builders.kt#L122-L144 What do you think?

File: kotlinx-coroutines-core/wasmWasi/src/internal/EventLoop.kt [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+private fun initializeSubscriptionPtr(allocator: MemoryAllocator): Pointer {
+    val ptrToSubscription = allocator.allocate(48)
+    //userdata
+    ptrToSubscription.storeLong(0)
+    //uint8_t tag;
+    (ptrToSubscription + 8).storeByte(0) //EVENTTYPE_CLOCK
+    //__wasi_clockid_t id;
+    (ptrToSubscription + 16).storeInt(CLOCKID_MONOTONIC) //CLOCKID_MONOTONIC
+    //__wasi_timestamp_t timeout;
+    //(ptrToSubscription + 24).storeLong(timeout)
```
Comment (dkhalanskyjb) [inline kotlinx-coroutines-core/wasmWasi/src/internal/EventLoop.kt:49]: Why is this commented out? The documentation of `allocate` states that it returns "uninitialized" memory; does it mean that the memory may contain garbage?
* If yes, we need to store the timeout;
* If not, what is the reason to explicitly write `0` in the other places in the same function? It looks like we have exactly one non-zero bit, the clock id.

Comment (igoriakovlev) [reply kotlinx-coroutines-core/wasmWasi/src/internal/EventLoop.kt:49]: It returns `uninitialized` memory and but waiting time loaded in `sleep` function.
All fields in this structure are the same between `sleep` calls but this `timeout` field. We load them once here and load the `timeout` in `sleep` function calls later.

**Comment by igoriakovlev** [comment]
I have rebased onto the master and merged all the AMEND commits by @dkhalanskyjb request.

**Comment by dkhalanskyjb** [comment]
Please do `./gradlew apiDump`; otherwise, the build fails.

File: kotlinx-coroutines-core/wasmWasi/src/internal/CoroutineRunner.kt [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+    val newContext = GlobalScope.newCoroutineContext(context)
+    val coroutine = object: AbstractCoroutine<Unit>(newContext, initParentJob = true, active = true) {}
+    coroutine.start(CoroutineStart.DEFAULT, coroutine, block)
+    runEventLoop()
```
Comment (dkhalanskyjb) [inline kotlinx-coroutines-core/wasmWasi/src/internal/CoroutineRunner.kt:11]: Let's take a look at this test (which shouldn't be added to the codebase, except maybe as `@Ignored`):
```kotlin
    @Test
    fun testHappilyHanging() = runTest(timeout = 1.seconds) {
        // launch { // uncomment for the edge case where the job completes but its children don't
        withContext(NonCancellable) {
            println("Start hanging forever")
            Channel<Unit>().receive()
        }
        // }
    }
```

On the JVM, this test hangs, as expected: there is nothing to obtain from an empty channel, and we ignore the timeout-induced cancellation requests. On Wasm/JS, it fails with "Unknown" after a couple of seconds (ok, weird, but at least the test doesn't pass). On JS, it fails due to the Mocha 2-second timeout.

On WASI, it just exits and reports success. When the event loop has nothing more to do, it just considers the coroutine as finished.

What we can do in this case is platform-specific; for this case, I propose us to do something like
```kotlin
    check(coroutine.isCompleted) { "Coroutine $coroutine did not complete, but the system reached quiescence") }
```
directly below this line.

**Review by dkhalanskyjb — CHANGES_REQUESTED** [review]

File: kotlinx-coroutines-core/wasmWasi/src/internal/CoroutineRunner.kt [hunk 10]
```diff
@@ -9,5 +9,6 @@ public fun runTestCoroutine(context: CoroutineContext, block: suspend CoroutineS
     val coroutine = object: AbstractCoroutine<Unit>(newContext, initParentJob = true, active = true) {}
     coroutine.start(CoroutineStart.DEFAULT, coroutine, block)
     runEventLoop()
-    if (coroutine.isCancelled) throw coroutine.getCancellationException().let { it.cause ?: it }
+    coroutine.getCompletionExceptionOrNull()?.let { throw it }
+    check(coroutine.isCompleted) { "Coroutine $coroutine did not complete, but the system reached quiescence" }
```
Comment (dkhalanskyjb) [inline kotlinx-coroutines-core/wasmWasi/src/internal/CoroutineRunner.kt:13]: Eh, no, it's better to do it in the opposite order. `getCompletionExceptionOrNull` itself throws exceptions when the coroutine is still active, but really obscure ones.

**Review by dkhalanskyjb — APPROVED** [review]
